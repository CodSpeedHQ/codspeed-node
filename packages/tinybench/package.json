{
  "name": "@codspeed/tinybench",
  "version": "0.0.0",
  "description": "tinybench compatibility layer for CodSpeed",
  "keywords": [
    "codspeed",
    "benchmark",
    "tinybench",
    "performance"
  ],
  "main": "dist/index.cjs.js",
  "module": "dist/index.es5.js",
  "types": "dist/index.d.ts",
  "files": [
    "dist"
  ],
  "author": "Arthur Pastel <arthur@codspeed.io>",
  "repository": "https://github.com/CodSpeedHQ/codspeed-node/tree/main/packages/tinybench",
  "homepage": "https://codspeed.io",
  "license": "MIT",
  "devDependencies": {
    "@types/find-up": "^4.0.0",
    "@types/stack-trace": "^0.0.30",
    "jest-mock-extended": "^3.0.1"
  },
  "dependencies": {
    "@codspeed/core": "workspace:*",
    "find-up": "^6.3.0",
    "stack-trace": "1.0.0-pre1",
    "tinybench": "^2.3.1"
  }
}
