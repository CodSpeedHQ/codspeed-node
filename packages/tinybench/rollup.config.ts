import { defineConfig } from "rollup";
import pkg from "./package.json" assert { type: "json" };
import { declarationsPlugin, jsPlugins } from "../../rollup.options";

const entrypoint = "src/index.ts";

export default defineConfig([
  {
    input: entrypoint,
    output: [
      {
        file: pkg.types,
        format: "es",
        sourcemap: true,
      },
    ],
    plugins: declarationsPlugin,
  },
  {
    input: entrypoint,
    output: [
      {
        file: pkg.main,
        format: "cjs",
        sourcemap: true,
      },
      { file: pkg.module, format: "es", sourcemap: true },
    ],
    plugins: jsPlugins(pkg.version),
    external: ["@codspeed/core"],
  },
]);
