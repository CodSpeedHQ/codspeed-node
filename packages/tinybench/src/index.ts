import measurement from "@codspeed/core";
import { get as getStackTrace } from "stack-trace";
import path, { dirname } from "path";
import { findUpSync, Options } from "find-up";
import { Bench } from "tinybench";

export function withCodSpeed(bench: Bench): Bench {
  if (!measurement.isInstrumented()) {
    return bench;
  }
  const callingFile = getCallingFile();
  bench.run = async () => {
    for (const task of bench.tasks) {
      const uri = callingFile + "::" + task.name;
      measurement.startInstrumentation();
      await task.fn();
      measurement.stopInstrumentation(uri);
    }
    return bench.tasks;
  };
  return bench;
}

function getCallingFile(): string {
  const stack = getStackTrace();
  const callingFile = stack[2].getFileName(); // [here, withCodSpeed, actual caller]
  const gitDir = getGitDir(callingFile);
  if (gitDir === undefined) {
    throw new Error("Could not find a git repository");
  }
  return path.relative(gitDir, callingFile);
}

function getGitDir(path: string): string | undefined {
  const dotGitPath = findUpSync(".git", {
    cwd: path,
    type: "directory",
  } as Options);
  return dotGitPath ? dirname(dotGitPath) : undefined;
}
