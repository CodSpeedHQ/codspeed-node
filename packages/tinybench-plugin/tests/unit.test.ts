import { mockDeep, mockReset } from "jest-mock-extended";
const mockCore = mockDeep<Measurement>();

import type { Measurement } from "@codspeed/core";
import { Bench } from "tinybench";
import { withCodSpeed } from "..";

jest.mock("@codspeed/core", () => ({
  ...jest.requireActual("@codspeed/core"),
  measurement: mockCore,
}));

beforeEach(() => {
  mockReset(mockCore);
  jest.clearAllMocks();
});

describe("Benchmark.Suite", () => {
  it("simple suite", async () => {
    mockCore.isInstrumented.mockReturnValue(false);
    const bench = withCodSpeed(new Bench({ time: 100 }));
    const onComplete = jest.fn();
    bench.add("RegExp", function () {
      /o/.test("Hello World!");
    });
    bench.getTask("RegExp")?.addEventListener("complete", onComplete);
    await bench.run();

    expect(onComplete).toHaveBeenCalled();
    expect(mockCore.startInstrumentation).not.toHaveBeenCalled();
    expect(mockCore.stopInstrumentation).not.toHaveBeenCalled();
  });
  it("check core methods are called", async () => {
    mockCore.isInstrumented.mockReturnValue(true);
    await withCodSpeed(new Bench())
      .add("RegExp", function () {
        /o/.test("Hello World!");
      })
      .run();
    expect(mockCore.startInstrumentation).toHaveBeenCalled();
    expect(mockCore.stopInstrumentation).toHaveBeenCalledWith(
      "packages/tinybench-plugin/tests/unit.test.ts::RegExp"
    );
  });
  it("check suite name is in the uri", async () => {
    mockCore.isInstrumented.mockReturnValue(true);
    await withCodSpeed(new Bench())
      .add("RegExp", function () {
        /o/.test("Hello World!");
      })
      .add("RegExp2", () => {
        /o/.test("Hello World!");
      })
      .run();
    expect(mockCore.stopInstrumentation).toHaveBeenCalledWith(
      "packages/tinybench-plugin/tests/unit.test.ts::RegExp"
    );
    expect(mockCore.stopInstrumentation).toHaveBeenCalledWith(
      "packages/tinybench-plugin/tests/unit.test.ts::RegExp2"
    );
  });
  it.each([true, false])(
    "check console output(instrumented=%p) ",
    async (instrumented) => {
      const logSpy = jest.spyOn(console, "log");
      const warnSpy = jest.spyOn(console, "warn");
      mockCore.isInstrumented.mockReturnValue(instrumented);
      await withCodSpeed(new Bench({ time: 100 }))
        .add("RegExp", function () {
          /o/.test("Hello World!");
        })
        .add("RegExp2", () => {
          /o/.test("Hello World!");
        })
        .run();
      expect({
        log: logSpy.mock.calls,
        warn: warnSpy.mock.calls,
      }).toMatchSnapshot();
    }
  );
});
