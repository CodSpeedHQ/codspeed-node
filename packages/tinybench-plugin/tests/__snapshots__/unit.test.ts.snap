// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Benchmark.Suite check console output(instrumented=false)  1`] = `
{
  "log": [],
  "warn": [
    [
      "[CodSpeed] 2 benches detected but no instrumentation found, falling back to tinybench",
    ],
  ],
}
`;

exports[`Benchmark.Suite check console output(instrumented=true)  1`] = `
{
  "log": [
    [
      "[CodSpeed] running with @codspeed/tinybench v1.0.1",
    ],
    [
      "    ✔ Measured packages/tinybench-plugin/tests/unit.test.ts::RegExp",
    ],
    [
      "    ✔ Measured packages/tinybench-plugin/tests/unit.test.ts::RegExp2",
    ],
    [
      "[CodSpeed] Done running 2 benches.",
    ],
  ],
  "warn": [],
}
`;
